### PR TITLE
4328 support confirm model for change asset address

### PR DESCRIPTION
**Pull request by NNagato**

### CODE REVIEW

**Comment by secmask** [comment]
@halink0803 please help to review this MR

File: v3/common/types.go [hunk 1]
```diff
@@ -290,3 +290,17 @@ type UpdateTradingPairEntry struct {
...
 	TradingPairs []UpdateTradingPairEntry `json:"trading_pairs" binding:"required,dive"`
 }
+
+// ChangeAssetAddressPending hold state of being changed address of asset and waiting for confirm to apply.
+type ChangeAssetAddressPending pendingObject
+
+// ChangeAssetAddressEntry present data to create a change asset address
+type ChangeAssetAddressEntry struct {
+	ID      uint64 `json:"id" binding:"required"`
+	Address string `json:"address" binding:"required"`
```
Comment (halink0803) [inline v3/common/types.go]: should we have binding isEthereumAddress here?

File: v3/http/asset_change_address.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+	id, err := s.storage.CreateChangeAssetAddress(changeAssetAddress)
+	if err != nil {
+		httputil.ResponseFailure(c, httputil.WithError(err))
+		return
+	}
+	httputil.ResponseSuccess(c, httputil.WithField("id", id))
+}
+
+func (s *Server) checkChangeAssetAddressParams(changeAssetAddressEntry common.ChangeAssetAddressEntry) error {
```
Comment (halink0803) [inline v3/http/asset_change_address.go]: This check could be done in binding. Reference: https://github.com/KyberNetwork/reserve-stats/blob/50f95c3a5354c843f9cc103b8b6182661fb8b4d8/lib/httputil/validators/validators.go#L17

File: v3/http/asset_change_address.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	}
+	return nil
+}
+
+func (s *Server) getChangeAssetAddress(c *gin.Context) {
+	var input struct {
+		ID uint64 `uri:"id" binding:"required"`
+	}
+	if err := c.ShouldBindUri(&input); err != nil {
+		log.Println(err)
```
Comment (halink0803) [inline v3/http/asset_change_address.go]: please using Printf with message.

File: v3/http/asset_change_address.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	}
+	httputil.ResponseSuccess(c, httputil.WithData(result))
+}
+
+func (s *Server) confirmChangeAssetAddress(c *gin.Context) {
+	var input struct {
+		ID uint64 `uri:"id" binding:"required"`
+	}
+	if err := c.ShouldBindUri(&input); err != nil {
+		log.Println(err)
```
Comment (halink0803) [inline v3/http/asset_change_address.go]: Please use Printf with message.

File: v3/http/asset_change_address.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	}
+	httputil.ResponseSuccess(c)
+}
+
+func (s *Server) rejectChangeAssetAddress(c *gin.Context) {
+	var input struct {
+		ID uint64 `uri:"id" binding:"required"`
+	}
+	if err := c.ShouldBindUri(&input); err != nil {
+		log.Println(err)
```
Comment (halink0803) [inline v3/http/asset_change_address.go]: Please use Printf with message detail.

**Comment by halink0803** [comment]
Please fix travis.

File: v3/http/asset_change_address.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+	for _, changeAssetAddressEntry := range changeAssetAddress.Assets {
+		if err := s.checkChangeAssetAddressParams(changeAssetAddressEntry); err != nil {
+			log.Println("error", err)
+			httputil.ResponseFailure(c, httputil.WithError(err))
+			return
+		}
+	}
+
+	log.Println("still run here")
```
Comment (secmask) [inline v3/http/asset_change_address.go]: consider more detail message or remove.

File: v3/common/types.go [hunk 7]
```diff
@@ -290,3 +290,17 @@ type UpdateTradingPairEntry struct {
 type CreateUpdateTradingPair struct {
 	TradingPairs []UpdateTradingPairEntry `json:"trading_pairs" binding:"required,dive"`
 }
+
+// ChangeAssetAddressPending hold state of being changed address of asset and waiting for confirm to apply.
+type ChangeAssetAddressPending pendingObject
```
Comment (secmask) [inline v3/common/types.go]: use CreateChangeAssetAddress for consistent.

File: v3/common/types.go (hunk 1 above)
Comment (NNagato) [reply v3/common/types.go]: fixed in https://github.com/KyberNetwork/reserve-data/pull/556/commits/e8fd05b739955a8907bdb22bf73d7e4d0f75bed9

File: v3/http/asset_change_address.go (hunk 2 above)
Comment (NNagato) [reply v3/http/asset_change_address.go]: fixed in https://github.com/KyberNetwork/reserve-data/pull/556/commits/e8fd05b739955a8907bdb22bf73d7e4d0f75bed9

File: v3/http/asset_change_address.go (hunk 3 above)
Comment (NNagato) [reply v3/http/asset_change_address.go]: fixed in https://github.com/KyberNetwork/reserve-data/pull/556/commits/e8fd05b739955a8907bdb22bf73d7e4d0f75bed9

File: v3/http/asset_change_address.go (hunk 4 above)
Comment (NNagato) [reply v3/http/asset_change_address.go]: fixed in https://github.com/KyberNetwork/reserve-data/pull/556/commits/e8fd05b739955a8907bdb22bf73d7e4d0f75bed9

File: v3/http/asset_change_address.go (hunk 5 above)
Comment (NNagato) [reply v3/http/asset_change_address.go]: fixed in https://github.com/KyberNetwork/reserve-data/pull/556/commits/e8fd05b739955a8907bdb22bf73d7e4d0f75bed9

File: v3/http/asset_change_address.go (hunk 6 above)
Comment (NNagato) [reply v3/http/asset_change_address.go]: removed in https://github.com/KyberNetwork/reserve-data/pull/556/commits/e8fd05b739955a8907bdb22bf73d7e4d0f75bed9

File: v3/common/types.go (hunk 7 above)
Comment (NNagato) [reply v3/common/types.go]: fixed in https://github.com/KyberNetwork/reserve-data/pull/556/commits/4b6d71eb5ee9e1f266f3773a1a97abf1537e25df

File: http/httputil/validators/validators.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+	"reflect"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/gin-gonic/gin/binding"
+	"gopkg.in/go-playground/validator.v8"
+)
+
+// isEthereumAddress is a validator.Func function that returns true if given field
+// is a valid Ethereum address.
+func isEthereumAddress(_ *validator.Validate, _ reflect.Value, _ reflect.Value,
```
Comment (halink0803) [inline http/httputil/validators/validators.go]: Just checked again with the code, this validator already available: https://github.com/KyberNetwork/reserve-data/blob/d9066aedbbebe814c5b40cc6f7f4931a37eb5f0f/v3/http/validator.go#L15

Comment (NNagato) [reply http/httputil/validators/validators.go]: fixed in https://github.com/KyberNetwork/reserve-data/pull/556/commits/185053be42e3e65ed550401c33b7089f11be0696

**Pull request merged**